core/src/main/java/org/apache/iceberg/TableScanContext.java

```diff
@@ -0,0 +1,189 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.util.Collection;
+import java.util.Map;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.Expressions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+
+/**
+ * Context object with optional arguments for a TableScan.
+ */
+final class TableScanContext {
+  private final Long snapshotId;
+  private final Expression rowFilter;
+  private final boolean ignoreResiduals;
+  private final boolean caseSensitive;
+  private final boolean colStats;
+  private final Collection<String> selectedColumns;
+  private final ImmutableMap<String, String> options;
+  private final Long fromSnapshotId;
+  private final Long toSnapshotId;
+
+  private TableScanContext(Long snapshotId, Expression rowFilter, boolean ignoreResiduals,
+                           boolean caseSensitive, boolean colStats, Collection<String> selectedColumns,
+                           ImmutableMap<String, String> options, Long fromSnapshotId, Long toSnapshotId) {
+    this.snapshotId = snapshotId;
+    this.rowFilter = rowFilter;
+    this.ignoreResiduals = ignoreResiduals;
+    this.caseSensitive = caseSensitive;
+    this.colStats = colStats;
+    this.selectedColumns = selectedColumns;
+    this.options = options;
+    this.fromSnapshotId = fromSnapshotId;
+    this.toSnapshotId = toSnapshotId;
+  }
+
+  Long snapshotId() {
+    return snapshotId;
+  }
+
+  Expression rowFilter() {
+    return rowFilter;
+  }
+
+  boolean ignoreResiduals() {
+    return ignoreResiduals;
+  }
+
+  boolean caseSensitive() {
+    return caseSensitive;
+  }
+
+  boolean colStats() {
+    return colStats;
+  }
+
+  Collection<String> selectedColumns() {
+    return selectedColumns;
+  }
+
+  ImmutableMap<String, String> options() {
+    return options;
+  }
+
+  Long fromSnapshotId() {
+    return fromSnapshotId;
+  }
+
+  Long toSnapshotId() {
+    return toSnapshotId;
+  }
+
+  TableScanContext copy() {
+    return TableScanContext.builder(this).build();
+  }
+
+  static Builder builder() {
+    return new Builder();
+  }
+
+  static Builder builder(TableScanContext other) {
+    return new Builder(other);
+  }
+
+  static class Builder {
```

Comment:
I think a builder makes this refactor more awkward because the context is always copied twice. Once in the builder, then once when `copy` is called on the builder's result. It would be simpler to keep the refinement pattern, where every method that alters the context returns a new context. Like this:

```java
  TableScanContext caseSensitive() {
    if (this.caseSensitive) {
      return this;
    } else {
      return new TableScanContext(..., true /* case sensitive */, ...);
    }
  }
```

Reply:
Yeah, that's another option. I can change it to use that pattern.